### PR TITLE
[FEATURE] Sign GUI

### PR DESCRIPTION
**Pull request by seailz**
[contributing]: https://github.com/Negative-Games/Framework/blob/main/CONTRIBUTING.md

## Pull Request Etiquette

 

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

 

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code)
- [x] Documentation
- [ ] Other: \_____ 

 

Closes Issue: #96 
## Description

Adds a Sign GUI

## Documentation
```java
import games.negative.framework.gui.sign.SignGUI;

public class SignGUITest extends SignGUI {

    public SignGUITest() {
        setLine(1, player -> "---------");
        setLine(2, player -> "Enter some epic text above!");
        setLine(3, player -> "(this is cool)");

        onClose((player, lines) -> {
            System.out.println("Player " + player.getName() + " closed the sign GUI.");
            System.out.println("The lines were: ");
            for (String line : lines) {
                System.out.println(line);
            }
        });

    };
}
```

### CODE REVIEW

**Comment by NegativeDev** [comment]
Has this been tested?

File: src/main/java/games/negative/framework/gui/sign/SignGUI.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import lombok.Setter;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+import java.util.ArrayList;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * Custom SignGUI inputs
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignGUI.java:15]:
```suggestion
 * Sign GUI
```

File: src/main/java/games/negative/framework/gui/sign/SignGUI.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * @param index The line number (0-3)
+     * @param line The line text
+     */
+    public void setLine(int index, Function<Player, String> line) {
+        if (index <= 3)
+            lines.set(index, new SignLine(line));
+    }
+
+    /**
+     * Opens the sign GUI
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignGUI.java:40]:
```suggestion
     * Opens the menu
```

File: src/main/java/games/negative/framework/gui/sign/SignGUI.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+     * @param player The player to open the GUI for
+     * @throws ProtocolLibNotInstalledException If ProtocolLib is not installed
+     */
+    public void open(Player player) throws ProtocolLibNotInstalledException {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignGUI.java:44]:
```suggestion
    public void open(@NotNull Player player) throws ProtocolLibNotInstalledException {
```

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * Opens the sign GUI for the given player.
+     *
+     * @param player  The player to open the GUI for.
+     * @param signGUI THe gui to open
+     * @author FrostedSnowman & Seailz
+     */
+    public static void open(Player player, SignGUI signGUI) {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:45]:
```suggestion
    public static void open(@NotNull Player player, @NotNull SignGUI signGUI) {
```

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        SignEditorPacket signEditorPacket = new SignEditorPacket();
+        PacketContainer signDataPacket = protocol.createPacket(PacketType.Play.Server.TILE_ENTITY_DATA);
+
+        signEditorPacket.setLocation(new BlockPosition(0, 0, 0));
+
+        player.sendBlockChange(new Location(player.getWorld(), 0, 0, 0), Material.SIGN, (byte) 0);
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:51]: This part of the code would only work in 1.8 and below, please consider cross-version compatibility.

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        signGUI.getLines().forEach(line -> {
+            lines.add(line.getText().apply(player));
+        });
+
+        for (String line : lines) {
+            nbt.put("Text" + (line + 1), lines.size() > lines.indexOf(line) ? String.format("{\"text\":\"%s\"}", Utils.color(line)) : "");
+        }
+
+        nbt.put("id", "minecraft:sign");
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:65]: This part of the code would only work in 1.8 and below, please consider cross-version compatibility.

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        });
+    }
+
+    /**
+     * Sends a packet with ProtocolLib
+     *
+     * @param player The player to send the packet to.
+     * @param packet The packet to send.
+     */
+    private static void sendPacket(Player player, PacketContainer packet) {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:104]:
```suggestion
    private static void sendPacket(@NotNull Player player, @NotNull PacketContainer packet) {
```

File: src/main/java/games/negative/framework/gui/sign/packets/SignEditorPacket.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    public static final PacketType TYPE =
+            PacketType.Play.Server.OPEN_SIGN_EDITOR;
+
+    public SignEditorPacket() {
+        super(new PacketContainer(TYPE), TYPE);
+        handle.getModifier().writeDefaults();
+    }
+
+    public SignEditorPacket(PacketContainer packet) {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/packets/SignEditorPacket.java:37]:
```suggestion
    public SignEditorPacket(@NotNull PacketContainer packet) {
```

File: src/main/java/games/negative/framework/gui/sign/packets/SignEditorPacket.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    public BlockPosition getLocation() {
+        return handle.getBlockPositionModifier().read(0);
+    }
+
+    /**
+     * Set Location.
+     *
+     * @param value - new value.
+     */
+    public void setLocation(BlockPosition value) {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/packets/SignEditorPacket.java:55]:
```suggestion
    public void setLocation(@NotNull BlockPosition value) {
```

File: src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Constructs a new strongly typed wrapper for the given packet.
+     *
+     * @param handle - handle to the raw packet data.
+     * @param type - the packet type.
+     */
+    protected AbstractPacket(PacketContainer handle, PacketType type) {
+        // Make sure we're given a valid packet
+        if (handle == null)
+            throw new IllegalArgumentException("Packet handle cannot be NULL.");
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java:44]:
```suggestion
    protected AbstractPacket(@NotNull PacketContainer handle, @NotNull PacketType type) {
```

File: src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        return handle;
+    }
+
+    /**
+     * Send the current packet to the given receiver.
+     *
+     * @param receiver - the receiver.
+     * @throws RuntimeException If the packet cannot be sent.
+     */
+    public void sendPacket(Player receiver) {
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java:67]:
```suggestion
    public void sendPacket(@NotNull Player receiver) {
```

File: src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    @Deprecated
+    public void recievePacket(Player sender) {
+        try {
+            ProtocolLibrary.getProtocolManager().recieveClientPacket(sender,
+                    getHandle());
+        } catch (Exception e) {
+            throw new RuntimeException("Cannot recieve packet.", e);
+        }
+    }
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java:99]: Why does this need to be here?

**Review by NegativeDev — CHANGES_REQUESTED** [review]

File: src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java (hunk 12 above)
Comment (seailz) [reply src/main/java/games/negative/framework/gui/sign/packets/general/AbstractPacket.java:99]: This is from a packet wrapper repository, will remove

File: src/main/java/games/negative/framework/gui/sign/SignManager.java (hunk 5 above)
Comment (seailz) [reply src/main/java/games/negative/framework/gui/sign/SignManager.java:52]: not sure how I'm supposed to do this? Material#getValue?

**Comment by seailz** [comment]
## TODO:
- [ ] Newer Version Support
- [ ] Newer Version Testing
- [ ] Fix Default Lines

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * Opens the sign GUI for the given player.
+     *
+     * @param player  The player to open the GUI for.
+     * @param signGUI THe gui to open
+     * @author FrostedSnowman & Seailz
+     */
+    public static void open(@NotNull Player player, @NotNull SignGUI signGUI) {
+        VersionChecker versionChecker = new VersionChecker();
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:58]: We already have a VersionChecker instance initialized, please do not initialize a new one. Please use 
```
VersionChecker.getInstance()
```

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        SignEditorPacket signEditorPacket = new SignEditorPacket();
+        PacketContainer signDataPacket = protocol.createPacket(PacketType.Play.Server.TILE_ENTITY_DATA);
+
+        setLocation(new BlockPosition((int) player.getLocation().getX(), (int) player.getLocation().getY(), (int) player.getLocation().getZ()));
+
+        if (versionChecker.isModern())
+            player.sendBlockChange(new Location(player.getWorld(), 0, 0, 0), Material.SIGN, (byte) 0);
+        else if (versionChecker.isLegacy())
+            player.sendBlockChange(new Location(player.getWorld(), 0, 0, 0), Material.valueOf("OAK_SIGN"), (byte) 0);
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:67]:
```suggestion
player.sendBlockChange(new Location(player.getWorld(), 0, 0, 0, versionChecker.isLegacy() ? Material.SIGN : Material.valueOf("OAK_SIGN"), (byte) 0));
```

File: src/main/java/games/negative/framework/gui/sign/SignManager.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+                lines.add(line.getText().apply(player));
+            else
+                lines.add("");
+        });
+
+
+
+        // String itemName = versionChecker.isModern() ? "minecraft:sign" : "minecraft:oak_sign";
+
+        signDataPacket.getBlockPositionModifier().write(0, location);
```
Comment (NegativeDev) [inline src/main/java/games/negative/framework/gui/sign/SignManager.java:84]:
```suggestion
        String itemName = versionChecker.isLegacy() ? "minecraft:sign" : "minecraft:oak_sign";
```

**Review by NegativeDev — CHANGES_REQUESTED** [review]

**Comment by seailz** [comment]
@ericlmao can this stay open? I'm gonna start trying to work on it again

**Comment by seailz** [comment]
> I'm gonna say no, I'll probably make my own implementation of it

bruh moment

**Comment by seailz** [comment]
let me do some testing with this, I think it may be working

**Pull request closed without merging**